lib/stud/interval.rb

```diff
@@ -1,30 +1,70 @@
 require "stud/task"
+require "thread"
+
 module Stud
+  STUD_STOP_REQUESTED = :stud_stop_requested
+
+  extend self
+
   # This implementation tries to keep clock more accurately.
   # Prior implementations still permitted skew, where as this one
   # will attempt to correct for skew.
   #
   # The execution patterns of this method should be that
   # the start time of 'block.call' should always be at time T*interval
-  def self.interval(time, opts = {}, &block)
+  def interval(time, opts = {}, &block)
     start = Time.now
     while true
-      break if Task.interrupted?
       if opts[:sleep_then_run]
         start = sleep_for_interval(time, start)
+        break if stop?
         block.call
       else
         block.call
         start = sleep_for_interval(time, start)
+        break if stop?
       end
     end # loop forever
   end # def interval
 
-  def self.sleep_for_interval(time, start)
+  # stop! instructs interval to stop and exit its execution loop before going to
+  # sleep between block executions.
+  # NOW the tricky part is: this is typically an operation that will be called
+  # from another thread than the thread running the interval loop in which case
+  # the target parameter must be set to the Thread object which is running the
+  # interval loop.
+  # Note that the stop logic is compatible with Stud::Task so if interval is run
+  # inside a Stud::Task, calling Stud::Task#stop! will stop the interval the same
+  # way as calling stop! on the interval itself.
+  # @param [Thread] the target thread to stop, defaut to Thread.current
+  def stop!(target = Thread.current)
+    target[STUD_STOP_REQUESTED] = true
+    target.wakeup
```

Comment:
blindly calling wakeup here could be a problem, if a task does some sleeps in its execution (for timing some transaction, for example) then it will abort that sleep and go to the actual Interval sleep. We need to know that we're in the `sleeping` part of the loop before calling wakeup